### ISSUE TITLE
Unable to include relative TT templates from within another template

### ISSUE THREAD

**Issue by abeverley** [PerlDancer/Dancer2#1093]
#1037 removed the default include path for TT templates. Whilst there were good reasons for this, it now means that by default a relative template cannot be included from another template. E.g. any application that was doing ```[% PROCESS another_template.tt %]``` will now be broken.

The simple fix is to define ```INCLUDE_PATH``` in the TT config. This is undesirable though, as it requires an absolute path.

Can I suggest as a solution that a new configuration option is added to the template_toolkit configuration to enable the default include path as ```$self->views``` so as to save hard-coding paths into the config?

If the above sounds okay, I will be happy to supply a PR.

Thanks,

Andy

**Comment by abeverley** [comment]
So, I'm going to try and summarise the issues and potential solutions:

1. Using ```set views => 'myviews'``` with a relative path did not work. This was because ```config->{views}``` was normally configured as an absolute path (built by Dancer on start-up). The config value was passed both to TT for its include path and also used to generate absolute path names passed to TT for each page render. When set to a relative path, Dancer was passing the constructed relative path filename, and TT was searching its (same) include path, leading to duplicate paths (e.g. ```myviews/myviews/index.tt```

2. The recent solution (#1037) fixes the above problem of a relative template directory, but has 2 problems: firstly the original subject of this issue (unable to include files within templates, or needing to include them with a different path, breaking existing applications), but also does not work anyway when an application is deployed using FastCGI (and possibly others?), as FastCGI-deployment means that the app does not find the templates unless the paths are passed as absolute values (I've not been able to work out exactly why this is, but is obviously something to do with working directories when deployed in that manner).

For the reasons outlined in the last part of para 2, I believe that absolute paths should always be passed to TT. Therefore, I can see only 2 solutions:

1. Revert the original patch (#1037) and always use ```set => views``` with an absolute path. The documentation already states this, although the example could be changed from ```set views => path(dirname(__FILE__), 'templates');``` to ```set views => config->{appdir}.'myviews'```

2. Revert the original patch (#1037) and provide a new function to set a relative template path. Of course, the existing ```set``` functionality can't be configured to be "intelligent" as it's a generic function to change configuration.

Incidentally, I think #955 also has problems, as it would still result in relative paths being used for TT, which has problems as described above.

**Comment by xsawyerx** [comment]
@veryrusty You're being assigned to this because you know way more than I do what's going on with this.

@abeverley, excellent write-up!

**Comment by abeverley** [comment]
@veryrusty - just to clarify my 2 comments, the first describes the problem, but has a bad proposed solution so ignore it. The second gives the real detail and some better proposals.

**Comment by racke** [comment]
On 01/06/2016 09:19 PM, Sawyer X wrote:
> @veryrusty <https://github.com/veryrusty> You're being assigned to this because you know way more than I do what's going on with this.
> 
> @abeverley <https://github.com/abeverley>, excellent write-up!
> 

My suggestion would be to revert the change causing the problem if we can't find a
better solution in the next few days.

Regards
	Racke

-- 
Perl and Dancer Development

Visit our Perl::Dancer conference 2015.
More information on https://www.perl.dance.

**Comment by veryrusty** [comment]
@racke reverting is going to break any apps relying on the patch for resolving #951.

Its all relative. #951 wanted relative paths for `views`. The patch in #1037 broke relative paths within templates (even `[% INCLUDE foo.tt %]` is relative to _something_). I'm going to revisit my approach in #955 over the coming weekend.

To be honest, I'm not concerned about some deployment options (like FastCGI) not supporting relative paths for views, as long as its possible to define an INCLUDE_PATH in the app's config and have that be searched _first_ by TT2 (which #955 didn't allow) - and that this is documented (probably within the deployment POD).

**Comment by veryrusty** [comment]
#1095 has a proposed fix, which is a revert of #1037 and inclusion of #955 with some extra goodness.

It adds a new config option
```
engines:
    template:
        template_toolkit:
            include_path: /absolute/path/to/templates
```

That prepends that path to the (generated) list for `INCLUDE_PATH` (yes, take care with the lower/upper case). That should allow relative views specified in development but the inclusion of a absolute path in production environments when the deployment choice doesn't play nicely with relative paths (/me stares at FastCGI).

**Comment by aka-raj** [comment]
So just to be clear, the proposed fix means that development env would revert to behaving exactly as it always did prior to 0.164/5, and only deployment under fastcgi would require include_path: /absolute/path/to/templates ? 

Any scope for setting an environment variable to the absolute path in the fastcgi config (eg apache2.conf) as I already do for PLACK_ENV and PERL5LIB ?

**Comment by veryrusty** [comment]
@aka-raj Yes. Use `$ENV{DANCER_VIEWS}` to set the default for the `views` setting.

**Comment by aka-raj** [comment]
Thanks. Proposed fix OK for me then.

**Comment by abeverley** [comment]
On Sat, 2016-01-09 at 06:27 -0800, aka-raj wrote:
> So just to be clear, the proposed fix means that development env would 
> revert to behaving exactly as it always did prior to 0.164/5, and only 
> deployment under fastcgi would require include_path: 
> /absolute/path/to/templates ?

It's better than that. fastcgi is exactly how it was too. You only need
to configure the absolute path within fastcgi if you are using a
relative ```views``` configuration, which was never supported before
anyway.

**Comment by aka-raj** [comment]
Getting better every clarification :)

**Comment by veryrusty** [comment]
Merged fix in #1095. :dancers:

@abeverley++ Thanks again for raising the issue and testing the proposed fix.
Thanks @aka-raj for your input too!
